Research/ResearchMotion/RSDMotionAudioSessionController.swift

```diff
@@ -0,0 +1,122 @@
+///
+/// - note: Using this audio session controller includes playing an audio file in the background in order to
+/// keep the app responsive to step transitions and voice commands. Without the audio playing in the background
+/// or the use of GPS, the app will stop responding to timing events.
+public final class RSDMotionAudioSessionController : NSObject, RSDAudioSessionController {
+    
+    public override init() {
+        super.init()
+        
+        // Load the audio file.
+        do {
+            let bundle = Bundle(for: RSDMotionAudioSessionController.self)
+            let url = bundle.url(forResource: "Silence", withExtension: "wav")!
+            self.audioPlayer = try AVAudioPlayer(contentsOf: url)
+            self.audioPlayer.numberOfLoops = -1
+            self.audioPlayer.prepareToPlay()
+        } catch let error {
+            debugPrint("Failed to open audio file. \(error)")
+        }
+    }
+    
+    /// The audio session is a shared pointer to the current audio session (if running). This is used to
+    /// allow background audio. Background audio is required in order for an active step to play sound
+    /// such as voice commands to a participant who make not be looking at their screen.
+    ///
+    /// For example, a "Walk and Balance" task that measures gait and balance by having the participant
+    /// walk back and forth followed by having them turn in a circle would require turning on background
+    /// audio in order to play spoken instructions even if the screen is locked before putting the phone
+    /// in the participant's pocket.
+    ///
+    /// - note: The application settings will need to include setting capabilities appropriate for
+    /// background audio if this feature is used.
+    ///
+    public private(set) var audioSession: AVAudioSession?
+    
+    /// The audio player that plays the background sound used to keep the motion sensors active. This is a
+    /// work-around to allow the task to continue running in the background without requiring GPS by instead
+    /// playing an audio file of silence.
+    ///
+    /// - note: As of this writing, speech-to-text using the `AVSpeechSynthesizer` will *not* run in the
+    /// background after 5 seconds and turning on background audio using `AVAudioSession` is not enough to
+    /// keep any timers running. syoung 05/21/2019
+    ///
+    public private(set) var audioPlayer: AVAudioPlayer!
+    
+    /// Start the background audio session if needed. This will look to see if `audioSession` is already started
+    /// and if not, will start a new session.
+    public func startAudioSessionIfNeeded() {
+        guard audioSession == nil else { return }
+        
+        // Start the background audio session
+        do {
+            let session = AVAudioSession.sharedInstance()
+            if #available(iOS 12.0, *) {
+                try session.setCategory(.playback, mode: .voicePrompt, options: .interruptSpokenAudioAndMixWithOthers)
```

Comment:
If we decided to replace the text-to-speech with recorded audio instructions, would this interrupt it with "Silence.wav"?

Reply:
In theory it should "mix" with other audio files, but this isn't really designed to be used where there is a recording.

For CRF, I am *not* planning on using this audio session controller because it isn't needed. B/c the stair step task uses an audio file, it will remain active, even in the lock screen.

The problem, I discovered, is that if you space out your voice prompts by more than 5 seconds, the app goes to sleep and no longer runs in the background.